netty/src/main/java/io/grpc/netty/NettyServer.java

```diff
@@ -309,20 +314,16 @@ public void operationComplete(ChannelFuture future) throws Exception {
                   public void operationComplete(ChannelFuture future) throws Exception {
                     channelz.addListenSocket(listenSocketStats);
                     synchronized (NettyServer.this) {
-                      listenSocketStatsList.add(listenSocketStats);
+                      socketStats.add(listenSocketStats);
```

Comment:
This is not thread-safe, as it can happen after `future.awaitUninterruptibly().isSuccess()` returns.

We had discussed needing to have another future that is completed right after this synchronized block, and pass that future instead in `bindFutures`. But I actually think it'd be easier to move this code (and listenSocketStats creation) to the `channelFutures.entrySet()` for loop directly. At that location, the entire contents of this `ChannelFutureListener` could be called directly and the synchronized block would be unnecessary. It'd be best to move `future.channel().closeFuture().addListener()` for `channelz.removeListenSocket()` to the for loop as well so they stay paired (but it would stay as a listener).

Reply:
Why synchronization block (L352) is not enough to ensure all write to the arraylist is sequentially done and the read is also up to date?

Reply:
Because this listener may be executed after that block.

Reply:
k, could you remind me why we originally have [this](https://github.com/grpc/grpc-java/blob/master/netty/src/main/java/io/grpc/netty/NettyServer.java#L271-L276) in eventloop?

Reply:
I don't think it is necessary. What is necessary is that it is executed after the bind completes. It appears it could have been simplified when https://github.com/grpc/grpc-java/commit/49c8bdb60ac8783eda098aa42375b837342edde3 was done, but wasn't noticed at the time (nor in https://github.com/grpc/grpc-java/commit/b8474d61c966f60cc0bad4e5abaca351e90afc9d)

Reply:
Thanks. I put it inline, but I am actually not hundred percent sure if `L341` is publish-safe. Reference assignment is fine, and because the list is modified in a single thread, it looks other thread would `get()` what this thread can see by the time unmodifiable view is created, which might be safe.

Reply:
It looks fine. Yes, the volatile write publishes any writes that occurred before it on the same thread.